### ISSUE TITLE
Hard to know what file is causing error

### ISSUE THREAD

**Issue by andoks** [safe-waters/docker-lock#89]
I am trying to introduce docker-lock into a relatively large git repository with a lot of different docker and compose files, but when errors are discovered that works when using docker/docker-compose (and therefore has passed by unnoticed before), it is hard to know what file the error located in.

Would it be possible to print the path of the file with the error?

An example of an error I have is a compose-file which has the service.target.publish port value specified as a string instead of a integer literal ("8888" instead of just 8888), this works fine with docker-compose, but errors out with docker-lock

**Comment by michaelperel** [comment]
@andoks Could you post an example of the compose-file that works with docker-compose but not with docker-lock?

**Comment by michaelperel** [comment]
@andoks great suggestion on file name. 

In general, `docker-lock` supports all versions 3.XX. 

Looking through the docker code, I see that it is supposed to be an integer.

```
type ServicePortConfig struct {
	Mode      string `yaml:",omitempty" json:"mode,omitempty"`
	Target    uint32 `yaml:",omitempty" json:"target,omitempty"`
	Published uint32 `yaml:",omitempty" json:"published,omitempty"`
	Protocol  string `yaml:",omitempty" json:"protocol,omitempty"`
}
```

However, they must be doing some kind of interpolation that is allowing the string to work. Currently looking into why that's the caes.

**Comment by michaelperel** [comment]
@andoks I have found the issue, and you may run into something like this in other scenarios too.

TLDR: `docker-lock` supports all versions 3.XX as well as `docker stack` does.

`docker-lock` uses the same code as the official go docker cli so that it can handle many edge cases. This cli provides a method to load `docker-compose` files, which `docker stack` uses. Specifically, both `docker-lock` and `docker stack` rely on [this method](https://github.com/docker/cli/blob/88c6089300a82d3373892adf6845a4fed1a4ba8d/cli/command/stack/loader/loader.go#L28).

Unfortunately, this method is a little bit stricter than the `docker-compose` python cli. So `docker-lock` will work wherever `docker stack` works, which [according to docker is all 3.XX versions](https://docs.docker.com/engine/swarm/stack-deploy/), but as we see, there are some slight differences.

To see how `docker stack` fails where docker-compose succeeds, try this file

```
version: "3.8"
services:
  db:
    image: couchdb
    ports:
     - published: "5984"
       target: "5984"
    deploy:
      restart_policy:
        condition: on-failure
```

If you run
```
docker stack deploy --compose-file docker-compose.yml test
```
It will fail because the ports are not integers.

If you run
```
docker-compose up
```
It will succeed.

Similarly, `docker lock generate` will fail if the ports are not integers.

As for your original request for error messages with paths, I will add them shortly in the future. Thanks!

**Comment by michaelperel** [comment]
@andoks I just merged in changes that should address your concerns about the path being printed when the parser fails. Could you let me know if any error messages are still unclear/getting in your way? Thanks!

**Comment by andoks** [comment]
@michaelperel: thanks! It now prints path for the error in compose-files which has the target set as a string instead of an integer. But after fixing that, I get the following error "'name' cannot be empty", which I do not know if is an error in a Dockerfile, a compose-file or stems from me calling docker lock in the wrong way somehow. Do you recognize the error? If it stems from something being wrong with a file it would really help if the path was printed in this instance too.

**Comment by michaelperel** [comment]
@andoks Thank you for for working through this, I really appreciate it.

I added file paths to what I thought the most common errors would be to try to make it expedient in solving your problem, but I will add it to all of the errors in the coming days.

For this specific error, you will get it when docker-lock parses a file but there is no image name. For instance in:
```
version: '3'
services:
  web:
    image: ''
```

**Comment by michaelperel** [comment]
@andoks The most recently merged PR should fix this, and looking through the code it seems to be the last place to me where you would get an opaque error message on generate.

**Comment by andoks** [comment]
@michaelperel my pleasure, but also thank _you_ for taking the time with this.

> For this specific error, you will get it when docker-lock parses a file but there is no image name. For instance in:

Well, then I understand the problem: we have "composition" compose-files, where we pass multiple compose files to docker-compose (e.g `docker-compose --file base-system.docker-compose.yml --file host-specific-options.docker-compose.yml`). The "host-specific-options.docker-compose.yml" file in this example may contain only the volumes and port bindings for all the service entries defined in "base-system.docker-compose.yml". Do note that the "addon" docker-compose.yml files _may_ contain image stanzas for certain services (or additional services altogether).

Making docker-lock work for this scenario is probably worth a new issues; but the short of it would be to have docker-lock not error out on service-definitions that does not contain an image stanza (and perhaps log or error out if the image stanza is present, but not a valid image)

Regarding the improved error message: I have verified that I now get the path printed as part of the error message. Thank you very much :+1:

**Comment by michaelperel** [comment]
Glad that the error messages are resolved -- will close this issue and look at the composition use case in the other issue!

Thanks so much for taking the time to go through this :)

**Comment by andoks** [comment]
@michaelperel thank _you_ for adding this :smile:
